### PR TITLE
Update k3s, containerd, kubelet config etc to make image more performant

### PR DESCRIPTION
**Pull request by aledbf**
**Changes:**
- Update k3s to v1.21.3
- Update containerd configuration:
Enable  `[plugins."io.containerd.grpc.v1.cri".registry]` to allow the use of self-signed SSL certificates

- Switch to [nerdctl](https://github.com/containerd/nerdctl) as the source of contained dependencies: (this simplifies the update)
  - nerdctl: v0.11.0
  - containerd: v1.5.4
  - runc: v1.0.1
  - CNI plugins: v0.9.1
  - CNI isolation plugin: v0.0.3
  - BuildKit: v0.9.0
  - Stargz Snapshotter: v0.7.0
  - imgcrypt: v1.1.1
  - RootlessKit: v0.14.2
  - slirp4netns: v1.1.11
  - fuse-overlayfs: v1.6
  - containerd-fuse-overlayfs: v1.0.3

- Use Ubuntu ppa to install the Linux kernel
- Add KubeletConfiguration and adjust defaults.
  - `"serializeImagePulls": false` -> from true
  - `"resolvConf": "/run/systemd/resolve/resolv.conf"` -> from /etc/resolv.conf - bypass Ubuntu local DNS (systemd-resolved)
  - `"registryBurst": 20` -> from 10
  - `"registryPullQPS": 15` -> from 5
  - `"imageGCHighThresholdPercent": 70` -> from 85
  - `"imageGCLowThresholdPercent": 50` -> from 80
- Update configuration of sysctl settings (mirror GCE defaults)

**Sources:**
- https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/
- https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/
- https://aws.amazon.com/es/premiumsupport/knowledge-center/eks-worker-nodes-image-cache/

### CODE REVIEW

File: setup.sh [hunk 1]
```diff
@@ -31,6 +31,7 @@ apt --no-install-recommends install -y \
   iproute2 \
   auditd \
   ethtool \
+  net-tools \
```
Comment (princerachit) [inline setup.sh:34]: should we pin these to some version?

File: setup.sh [hunk 2]
```diff
@@ -46,27 +47,26 @@ blacklist dccp
...
-dpkg --force-all -i ./*.deb
-rm ./*.deb
+# Install linux kernel 5.13
+add-apt-repository -y ppa:tuxinvader/lts-mainline
+apt-get update
+apt-get install -y linux-generic-5.13
 
 # Configure grub
-echo "GRUB_GFXPAYLOAD_LINUX=keep" >> /etc/default/grub
+#echo "GRUB_GFXPAYLOAD_LINUX=keep" >> /etc/default/grub
```
Comment (princerachit) [inline setup.sh:56]: If we don't plan to uncomment these then let us get rid of them

File: sysctl/11-network-security.conf [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# Turn on SYN-flood protections.  Starting with 2.6.26, there is no loss
```
Comment (princerachit) [inline sysctl/11-network-security.conf:1]: Can add the link to the page that you referred to create all the conf files?

File: setup.sh [hunk 4]
```diff
@@ -46,27 +47,26 @@ blacklist dccp
...
-curl -sSL -o linux-headers.deb        https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.12.5/amd64/linux-headers-5.12.5-051205-generic_5.12.5-051205.202105190541_amd64.deb
-curl -sSL -o linux-headers_all.deb    https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.12.5/amd64/linux-headers-5.12.5-051205_5.12.5-051205.202105190541_all.deb
-curl -sSL -o linux-image-unsigned.deb https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.12.5/amd64/linux-image-unsigned-5.12.5-051205-generic_5.12.5-051205.202105190541_amd64.deb
-curl -sSL -o linux-modules.deb        https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.12.5/amd64/linux-modules-5.12.5-051205-generic_5.12.5-051205.202105190541_amd64.deb
-dpkg --force-all -i ./*.deb
-rm ./*.deb
+# Install linux kernel 5.13
+add-apt-repository -y ppa:tuxinvader/lts-mainline
+apt-get update
+apt-get install -y linux-generic-5.13
```
Comment (princerachit) [inline setup.sh:53]: Maybe a comment explaining that this can get outdated and we need to keep up with the version in upstream would help

**Review by princerachit — COMMENTED** [review]
PTAL at the comments

File: setup.sh (hunk 2 above)
Comment (aledbf) [reply setup.sh]: done

File: setup.sh (hunk 1 above)
Comment (aledbf) [reply setup.sh:34]: No. These are ubuntu packages, we should always use the latest version during a VM image build

File: sysctl/11-network-security.conf (hunk 3 above)
Comment (aledbf) [reply sysctl/11-network-security.conf:1]: These files where extracted from a GKE VM

File: setup.sh (hunk 1 above)
Comment (princerachit) [reply setup.sh:34]: Latest versions are susceptible to breaking changes. We have seen this problem in the past when we did not pin the versions of some npm/node modules and ended up with breaking changes.

When things in dependencies go wrong it is extremely hard to figure out root cause. WDYT?

**Pull request merged**